tool/smarts/src/test/java/org/openscience/cdk/isomorphism/SmartsExprReadTest.java

```diff
@@ -0,0 +1,901 @@
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.assertTrue;
+import static org.openscience.cdk.isomorphism.matchers.Expr.Type.*;
+
+public class SmartsExprReadTest {
+
+    static Expr expr(Expr.Type type) {
+        return new Expr(type);
+    }
+
+    static Expr expr(Expr.Type type, int val) {
+        return new Expr(type, val);
+    }
+
+    static Expr and(Expr a, Expr b) {
+        return new Expr(AND, a, b);
+    }
+
+    static Expr or(Expr a, Expr b) {
+        return new Expr(OR, a, b);
+    }
+
+    private static Expr getAtomExpr(IAtom atom) {
+        return ((QueryAtom) AtomRef.deref(atom)).getExpression();
+    }
+
+    private static Expr getBondExpr(IBond bond) {
+        return ((QueryBond) BondRef.deref(bond)).getExpression();
+    }
+
+    static Expr getAtomExpr(String sma, int flav) {
+        IAtomContainer mol = new AtomContainer();
+        assertTrue(Smarts.parse(mol, sma, flav));
+        return getAtomExpr(mol.getAtom(0));
+    }
+
+    static Expr getAtomExpr(String sma) {
+        return getAtomExpr(sma, Smarts.FLAVOR_LOOSE);
+    }
+
+    static Expr getBondExpr(String sma, int flav) {
+        IAtomContainer mol = new AtomContainer();
+        assertTrue(Smarts.parse(mol, sma, flav));
+        return getBondExpr(mol.getBond(0));
+    }
+
+    static Expr getBondExpr(String sma) {
+        return getBondExpr(sma, Smarts.FLAVOR_LOOSE);
+    }
+
+    @Test
+    public void trailingOperator() {
+        IAtomContainer mol = new AtomContainer();
+        assertFalse(Smarts.parse(mol, "[a#6,]"));
+        assertFalse(Smarts.parse(mol, "[a#6;]"));
+        assertFalse(Smarts.parse(mol, "[a#6&]"));
+        assertFalse(Smarts.parse(mol, "[a#6!]"));
+    }
+
+    @Test
+    public void leadingOperator() {
+        IAtomContainer mol = new AtomContainer();
+        assertFalse(Smarts.parse(mol, "[,a#6]"));
+        assertFalse(Smarts.parse(mol, "[;a#6]"));
+        assertFalse(Smarts.parse(mol, "[&a#6]"));
+        assertTrue(Smarts.parse(mol, "[!a#6]"));
+    }
+
+    @Test
+    public void trailingBondOperator() {
+        IAtomContainer mol = new AtomContainer();
+        assertFalse(Smarts.parse(mol, "*-,*"));
+        assertFalse(Smarts.parse(mol, "*-;*"));
+        assertFalse(Smarts.parse(mol, "*-&*"));
+        assertFalse(Smarts.parse(mol, "*-!*"));
+    }
+
+    @Test
+    public void leadingBondOperator() {
+        IAtomContainer mol = new AtomContainer();
+        assertFalse(Smarts.parse(mol, "*,-*"));
+        assertFalse(Smarts.parse(mol, "*;-*"));
+        assertFalse(Smarts.parse(mol, "*&-*"));
+        assertTrue(Smarts.parse(mol, "*!-*"));
+    }
+
+    @Test
+    public void opPrecedence1() {
+        IAtomContainer mol = new AtomContainer();
+        assertTrue(Smarts.parse(mol, "[a#6,a#7]"));
+        Expr actual = getAtomExpr(mol.getAtom(0));
```

Comment:
What is the meaning of getAtomExpr(null)?

Reply:
(Or does Smarts.parse() not do what I think it does... /me is scrolling back...)

Reply:
I don't know what you're asking, getAtomExpr is just a helper function for the tests

Reply:
The question was basically, link 132 does not create a molecule, just an empty container... so, unless line 133 is adding atoms and bonds to that container, how is it going to map anything at all in line 134? So, is Smarts.parse() in line 133 creating a molecule? Or otherwise, why is mol.getAtom(0) in line 134 not returning null?

Reply:
Parse adds atoms/bonds to the molecule, doesn't matter if it's a query mol or not :-). Note I have an assertTrue, if the SMARTS is bad it, no atoms will be added but it returns false and so stops there.

Reply:
Ah, ok, clear! Thanks for clarifying this. I wasn't entirely sure what it was doing, but searching a substructure in a query substructure to test things is something I can certainly live with :)

(Plz ping me when you consider the patch final.)